<launch>
    <arg name="local_frame" default="base_link"/>
    <arg name="global_frame" default="odom"/>

    <arg name="omp_cores" default="8"/>
    <arg name="image_width" default="360"/>

    <arg name="voxel_size" default="0.2"/>
    <arg name="scan_radius" default="4.0"/>

    <arg name="use_imm" default="true"/>
    
    <node name="front_end_node" pkg="smat" type="rimg_front_end_node">
        <param name="local_frame" value="$(arg local_frame)"/>
        <param name="global_frame" value="$(arg global_frame)"/>

        <param name="omp_cores" value="$(arg omp_cores)"/>
        
        <param name="voxel_size" value="$(arg voxel_size)"/>
        <param name="sensor_height" value="0.4"/>
        <param name="cluster_size" value="10"/>
        <param name="max_depth" value="15.0"/>

        <!-- lidar param -->
        <rosparam file="$(find smat)/config/lidar_sim.yaml" command="load"/>

        <!-- Track param -->
        <rosparam file="$(find smat)/launch/params/nnt.yaml" command="load" />

        <!-- General IMM filter parameters -->
        <rosparam file="$(find smat)/launch/params/imm_CT_CV.yaml" command="load" if="$(arg use_imm)"/>
        
        <!-- Default Kalman filter parameters -->
        <rosparam file="$(find smat)/launch/params/ekf_default.yaml" command="load"/>
        
        <remap from="/points" to= "/os_cloud_node/points"/>
        <remap from="/init_map" to="/Laser_map"/>
        <remap from="/static_map" to="/map_merge_node/static_map"/>
    </node>

    <node name="back_end_node" pkg="smat" type="voi_back_end_node" output='screen'>
        <param name="local_frame" value="$(arg local_frame)"/>
        <param name="global_frame" value="$(arg global_frame)"/>
        <param name="voxel_size" value="$(arg voxel_size)"/>
        
        <param name="scan_radius" value="$(arg scan_radius)"/>
        <param name="max_nearest_size" value="80"/>
        <param name="downsampling_flag" value="false"/>

        <param name="omp_cores" value="$(arg omp_cores)"/>
        <param name="image_width" value="$(arg image_width)"/>

        <rosparam file="$(find smat)/config/lidar_sim.yaml" command="load"/>
    </node>

    <node name="map_merge_node" pkg="smat" type="map_merger_node">
        <param name="global_frame" value="$(arg global_frame)"/>
        <param name="output_static_map_path" value="/home/bwshen"/>

        <param name="omp_cores" value="$(arg omp_cores)"/>
        <param name="image_width" value="$(arg image_width)"/>

        <param name="merge_dist" value="$(arg scan_radius)"/>
        <param name="voxel_size" value="$(arg voxel_size)"/>

        <rosparam file="$(find smat)/config/lidar_sim.yaml" command="load"/>
    </node>

</launch>